frameworks.qmd

```diff
@@ -1,119 +1,1785 @@
+TensorFlow and PyTorch simplify applying random augmentations each epoch
+by integrating into the data pipeline.By programmatically increasing
+variation in the training data distribution, augmentations reduce
+overfitting and improve model generalization.
+
+Many frameworks make it easy to integrate augmentations into the data
+pipeline so they are applied on-the-fly each epoch. Together, performant
+data loaders and extensive augmentations enable practitioners to feed
+massive, varied datasets to neural networks efficiently. Hands-off data
+pipelines represent a significant improvement in usability and
+productivity. They allow developers to focus more on model architecture
+and less on data wrangling when training deep learning models.
+
+### Optimization Algorithms
+
+Training a neural network is fundamentally an iterative process that
+seeks to minimize a loss function. At its core, the goal is to fine-tune
+the model weights and parameters to produce predictions as close as
+possible to the true target labels. Machine learning frameworks have
+greatly streamlined this process by offering extensive support in three
+critical areas: loss functions, optimization algorithms, and
+regularization techniques.
+
+Loss Functions are useful to quantify the difference between the
+model\'s predictions and the true values. Different datasets require a
+different loss function to perform properly, as the loss function tells
+the computer the "objective" for it to aim to. Commonly used loss
+functions are Mean Squared Error (MSE) for regression tasks and
+Cross-Entropy Loss for classification tasks.
+
+Once the loss is computed, we need methods to adjust the model\'s
+parameters to reduce this loss or error during the training process. To
+do so, current frameworks use a gradient based approach, where it
+computes how much changes tuning the weights in a certain way changes
+the value of the loss function. Knowing this gradient, the model moves
+in the direction that reduces the gradient. There are many challenges
+associated with this, however, primarily stemming from the fact that the
+optimization problem is not convex, making it very easy to solve, and
+more details about this will come in the AI training sections. Modern
+frameworks come equipped with efficient implementations of several
+optimization algorithms, many of which are variants of gradient descent
+algorithms with stochastic methods and adaptive learning rates. More
+information can be found in the AI training section.
+
+Last but not least, overly complex models tend to overfit, meaning they
+perform well on the training data but fail to generalize to new, unseen
+data (see Overfitting). To counteract this, regularization methods are
+employed to penalize model complexity and encourage it to learn simpler
+patterns. Dropout for instance randomly sets a fraction of input units
+to 0 at each update during training, which helps prevent overfitting.
```

Comment:
Since the paragraph explains overfitting, wouldn't it also be helpful to touch on underfitting, and in what situations either case may be encountered? I think it would be informative to add examples with potential graphs to illustrate this point.